pycbc/io/live.py

```diff
@@ -373,27 +375,40 @@ def upload(self, fname, gracedb_server=None, testing=True,
         # as GraceDB operations can fail later
         self.save(fname)
 
+        # hardware injections need to be maked with the INJ tag
+        if self.is_hardware_injection:
+            labels = (labels or []) + ['INJ']
+
+        # connect to GraceDB if we are not reusing a connection
+        if not hasattr(self, 'gracedb'):
+            logging.info('Connecting to GraceDB')
+            gdbargs = {'reload_certificate': True, 'reload_buffer': 300}
+            if gracedb_server:
+                gdbargs['service_url'] = gracedb_server
+            try:
+                from ligo.gracedb.rest import GraceDb
+                self.gracedb = GraceDb(**gdbargs)
+            except Exception as exc:
+                logging.error('Failed to create GraceDB client')
+                logging.error(exc)
+
+        # create GraceDB event
+        logging.info('Uploading %s to GraceDB', fname)
+        group = 'Test' if testing else 'CBC'
         gid = None
         try:
-            if not hasattr(self, 'gracedb'):
-                from ligo.gracedb.rest import GraceDb
-                gdbargs = {'reload_certificate': True, 'reload_buffer': 300}
-                self.gracedb = GraceDb(gracedb_server, **gdbargs) \
-                    if gracedb_server else GraceDb(**gdbargs)
-            # create GraceDB event
-            group = 'Test' if testing else 'CBC'
-            r = self.gracedb.create_event(group, "pycbc", fname, search).json()
-            gid = r["graceid"]
+            response = self.gracedb.create_event(
```

Comment:
why are you getting rid of the case where self.gracedb was not defined? I guess it is harder to diagnose/debug if the creation of the gdb session can happen in two different places .. anyway, if we can reach this point and self.gracedb does not exist, we might want to produce an exception earlier on in this function